### ISSUE TITLE
Allow models to be revisioned, ordered, and editied through the admin without appearing in the explorer

### ISSUE THREAD

**Issue by alexgleason** [wagtail/wagtail#2270]
I don't think it makes sense to think about all manageable models in terms of pages.

Wagtailmodeladmin gets us most of the way there. But what we don't have, is the ability to create "things" in Wagtail that are also version controlled and rearrangeable. The workaround is to just let "things" be pages. But then you ​_have_​ to find a place for them in the Explorer. I don't want that, though. I want them to be separate.

The problem is that pages are not what a website is comprised of, they're how the website is delivered. This can change with technology -- an API for instance won't deliver content in terms of pages. It can also be confusing for content editors:

If a content editor wants to add a new team member, they shouldn't have to go under "about > team". They can click "team" in the left sidebar and add one right there. It could even have a cute "person" icon. How much more straightforward could it be?

I would rather not use "PersonPage"s for the sites I build. It's just a Person.

Okay, so how about displaying the content? Ideally, with blocks. Every aspect of the page body is built by blocks. Want a listing of people? Add a "Team Block" to your body StreamField. The team block is just a placeholder block in the admin, maybe with some options. Then on the frontend it displays the team listing. This way you can create a StandardPage called "Team" and add a team block to it. Or for more granular control, have different StreamFields for different page types.

Need team members to have their own URLs? You could hard-code the section URL in Django, like: `/person/$SLUG`, where the slug is dynamic. This is fine for "things." [Cool URLs don't change](https://www.w3.org/Provider/Style/URI), anyway.

What do you think? I think this could simplify things a lot. All we need to make this happen now is a base class to inherit from, similar to Page, but that doesn't appear in the Explorer. The admin side can be handled by Wagtailmodeladmin.

**Comment by heymonkeyriot** [comment]
I think this would be a great improvement. At the moment Wagtail websites are very well positioned to creating standard hierarchical tree websites, but it runs in to difficulties when it needs to have any other form (e.g. multi-hierarchy, hub and spoke or bento box style navigation flows)

I think it would also be likely to feel more intuitive to the user e.g. if a user creates a model they can see it immediately within the admin UI without needing to first create a new page (a little like snippets) And as you say overall simpler.

There's a few concerns/issues I could see

 - How do we deal with backwards compatibility? Could it be that a site could be set up as either hierarchical (as currently) or non-hierarchical (as per the above proposal)?
- Is there anyway we could see that existing sites could be ported to a new navigation pattern?
- How would menus be defined? Currently we have the `show in menu` button. Would that still be there? Or would there be some sort of menu creator? That feels like it may be adding greater flexibility at the cost of simplicity maybe? Would it also go against the premise of Wagtail being designed for developers to develop rather than editors to edit in the UI?
- How would we show hierarchy between items. Currently it's elementary to show a linear hierarchy (e.g. district -> schools -> teachers -> subjects) This feels like it would add complexity there. Could we still define those hierarchies via the UI? Defining parents perhaps?

**Comment by alexgleason** [comment]
Thanks Edd. :smile: 

I left my thoughts on implementation kinda ambiguous because I know it could be implemented a number of different ways. I'd like everyone's thoughts on it.

For instance:

* Introduce a new model like Page (but _not_ Page) to inherit from when making "things." Maybe we could call it Thing. But I'm sure there's a better name. Some CMS's call this a "Channel."
* Adjust the Page model so that you can configure subclasses to not appear in the Explorer.

> How do we deal with backwards compatibility?

What I'm proposing is to _add_ something on top of what we have, not remove or change anything. This shouldn't cause any problems with backwards compatibility. Old sites will work like they always have. Leave pages alone; add a concept of "things." The main difference is a mental one. It gives developers the ability to implement models in a more meaningful way.

> How would menus be defined?

Good question. Presumably if something is a "thing," it wouldn't be a menu item, so it's not necessary. Let me know if anyone has a counter argument.

> How would we show hierarchy between items
> (e.g. district -> schools -> teachers -> subjects)

Approach the whole thing differently. Instead of making a TeacherPage a child of a SchoolPage, you have Teachers and Schools on the same level. Teacher has a ForeignKey `school` field to set the school the teacher belongs to. How we display this in the admin is a reasonable concern, as there's no way to filter entries by their fields yet. It would display all the teachers for every school. Wasn't there an issue raise about this? I know Wharton brought it up regarding their blog. Either way I think it's a separate problem to solve.

Nesting things categorically by subpages only works in a 1-to-1 relationship scenario anyway. The only reason you put a teacher under a school is because teachers can only have _one_ school. What do you do in a many-to-many situation? The pages paradigm just doesn't allow it.

Instead of `/schools/harvard/edd/compsci` you would have:

* `/school/harvard`
* `/teacher/edd`
* `/study/compsci`

**Comment by davecranwell** [comment]
See also: https://github.com/torchbox/wagtail/issues/834 which in effect replicates a Drupal-like navigable, flat hierarchy of "content" (not "Pages"). I'm not familiar with "Channels" but I'm assuming they're as suitably vague as Drupal's "Node" concept, which could be a page, or an image, or anything really. Each Drupal Node is based on a "Content Type" which is practically the same as Django's content types - they're just a database table of things. Whether they are accessible via a URL in a tree-like structure is incidental.

**Comment by alexgleason** [comment]
I'm unfamiliar with Drupal but I expect that "nodes" are the same concept.

The only difference between a channel and a pure Django model is that a channel has entries that are versioned and can be rearranged.

**Comment by tmsndrs** [comment]
This feels very pertinent to me. There are lots of industry conversations going on about decoupling of content from form (hierarchy and presentation) and headless CMS. One of the issues I have with Wagtail is that it often encourages editors to think of the content they are creating in terms a specific presentational instance they have in mind (desktop). Even in the most basic contexts of RWD this isn't faithful to reality and will become ever truer as content is used in more exciting ways (APIs, personalisation, IOT etc). 

I haven't had a chance to check all of these out but here are some resources that talk about this issue. 
http://shoptalkshow.com/episodes/204-matt-dennewitz-jeff-eaton/ 
https://www.youtube.com/watch?v=QeAKaFccNWg
http://buytaert.net/the-future-of-decoupled-drupal (about Drupal but relevant)
https://pantheon.io/decoupled-cms

This obviously isn't something to take lightly and we would need to make sure we're not creating complicated workflows for editors. 

We do already have the concept of abstract objects with 'Snippets' but their recognised use case in more basic.

**Comment by alexgleason** [comment]
Agreed! Thanks for sharing @tmsndrs 

Snippets are close, but afaik can't be versioned or rearranged like pages can. Also I don't think they were intended to be used the way most of us use them.

**Comment by tomdyson** [comment]
Thanks for the feedback, everyone.

Note that many developers are already using Wagtail + wagtailmodeladmin (which can use any Django model, not just snippets) to achieve these goals, including fully-fledged headless CMSs. Revisions and moderation are the missing piece, and there's a lot of discussion about how to make this work at sprint many of the core devs are currently attending in Cape Town.

**Comment by alexgleason** [comment]
By the way, here is an example of a "StreamField" from a competing (non-free, PHP) CMS where I've implemented a "Team Listing Block" as I described before. It's the block on the bottom, but I'm showing the whole field for context.

![screenshot from 2016-02-24 04 51 57](https://cloud.githubusercontent.com/assets/3639540/13281548/7b3dc22e-dab2-11e5-8ae2-d6843b813935.png)

I've organized the block selector by type as well, where generated content is placed under widgets:

![screenshot from 2016-02-24 04 52 08](https://cloud.githubusercontent.com/assets/3639540/13281597/b703f99a-dab2-11e5-8684-173d881e7415.png)

Each of these "widgets" show up as placeholders in the admin. But on the front-end, they pull in data and display them. Here's what the team member block looks like rendered. The text at the top is the "story" block in the previous screenshot, and the accordion listing of people is the team member block.

![screenshot from 2016-02-24 04 55 32](https://cloud.githubusercontent.com/assets/3639540/13281623/e12adbda-dab2-11e5-8f9d-15912978acdb.png)

And you manage them like this:

![screenshot from 2016-02-24 04 58 14](https://cloud.githubusercontent.com/assets/3639540/13281689/37e474ae-dab3-11e5-92cf-2fae254a5661.png)

Then clicking one of those takes you to:

![screenshot from 2016-02-24 04 57 44](https://cloud.githubusercontent.com/assets/3639540/13281691/3b009e10-dab3-11e5-839a-88afff3cc1d2.png)

**Comment by heymonkeyriot** [comment]
I think what Alex is describing is slightly different to Drupal's content types, at least in the way that I think he's thinking they'll be applied within the UI. It feels closer to Wordpress's [custom post types](https://codex.wordpress.org/Post_Types) from my experience of those CMSs.

I think @tomdyson is correct that we're close to being at the point of Alex's suggestion with wagtailmodeladmin, and great to hear work is happening with it in South Africa! I wonder if wagtailmodeladmin could be something that's included within core rather than being a module to add?

Thanks for sharing the links @tmsndrs. Also agree with all your points.

**Comment by tomdyson** [comment]
> I wonder if wagtailmodeladmin could be something that's included within core rather than being a module to add?

We've been working towards this for 5 months :) See https://github.com/torchbox/wagtail/pull/1829. It's currently missing tests. There's a general sense that we need to accelerate this, as it's clearly a very important feature for many users.

**Comment by alexgleason** [comment]
@heymonkeyriot related: #1657 #1829

Also just remembed this very old Google Groups post I made similar to this topic but before I had as developed of an idea about it. https://groups.google.com/forum/#!msg/wagtail-developers/uqwztnHU6QY/JsRgr2UwAAAJ

**Comment by heymonkeyriot** [comment]
Sorry Tom, I'm late to the party as always :)

**Comment by tmsndrs** [comment]
This all sounds really interesting.

I think the point of listings is interesting. Currently the standard implementations I've seen has listings treated in the same way as any other content type when often they are very different from content types that have common content attributes across multiple instances of that particular content type. I guess listings have a more inherent relationship with a specific presentation of a dataset. In practise this results in weird content type field sets in Wagtail where some content types only provide editors with a 'title' field and something like an 'intro' field with no clear context of what that listing is doing. 

I would be really interested to learn more about how others are currently working on these types of issues with Wagtail. I'm really coming to this from a content modelling and IA perspective so keen to learn from dev insights.

**Comment by gasman** [comment]
Hi @alexgleason (et al) - great suggestion!

To elaborate on the discussions at Cape Town - it's becoming clear that for basically every distinguishing feature of the Page model (revisions, moderation workflow, having an HTML rendering, having a slug, translatability (soon)...) and equally Image / Document (belonging to collections, being insertable into rich text...), there's a plausible use-case for being able to attach that behaviour to arbitrary models (whether that's through snippets, or modeladmin). The logical conclusion is that Wagtail would implement a laundry list of model "traits" such as the above, and all of the models available in the admin, both core and custom, would incorporate some combination of those traits.

Moving to this kind of fully generic model would be a large and complex piece of work, especially if we're going to preserve the highly-tuned UI that we've got at the moment. In a way, we have the same endgame as Django CMS, but from different directions... Django CMS has taken django-admin and attached CMS functionality to it, whereas we're starting from a page-oriented CMS and genericising it into an alternative django-admin (but one that's highly optimised to the needs of content editors).

**Comment by ababic** [comment]
Hey guys! What @alexgleason is suggesting sounds very similar to how mezzanine (another, popular Django-powered CMS) arrange their core models. Lots of mixins that you can pick and choose from. Most complete being Displayable (almost a Page, but not part of the page tree) and Page:

https://github.com/stephenmcd/mezzanine/blob/master/mezzanine/core/models.py

While I was always fond of mezzanine's approach when working with it, I've grown to prefer the way wagtail makes you think about everything that has it's own front-end 'page' as a Page. 

Maybe it's just because of the nature of projects we've been using it for (website with lots of pages), but clients seem to feel very comfortable with the idea that every 'page' they see on the website's front-end is a Page in the CMS, irrespective of what it represents. And I have to say, it has it's benefits for development, too. Whether the Thing's end-point is defined in independently-defined URLs, or by where it lives in the tree, it still needs to be defined somehow in most cases. At least if it's the tree controlling the routing, the routing can handled consistently and dynamically. For example, if I have a news app with IndexPage and ArticlePage models, I can happily have 20 IndexPages dotted around my site if I really wanted to, each with it's own ArticlePages beneath it. And they can work perfectly fine without the need for custom-defined URLs (as long as each IndexPage is coded in a way as to only list its own children). And, if the client decided to delete a listing page (or all of them), or wanted to move some around to different places in the overall structure, that would be absolutely fine. I don't ever have to worry about maintaining a hardcoded list of URLs to match.

With wagtailmodeladmin enabling us to bypass the Explorer views to easily add an manage specific types of content, we don't find the 'everything is a Page' concept limits us or clients in any way, in terms of content management or development. Sure, it's taken some getting used to, and it's not a strictly 'conventional' approach for Django CMSs, but it is logical. Even in a less traditional website approach (where 'content' is served via APIs), I feel like I'd struggle to think of something that had an end-point, but didn't logically fit into a tree somewhere.... or at least, couldn't be represented with a simpler (possibly a snippet) model and served separately.

I guess what I'm saying is: Providing some kind of mixin of some kind to let non-page models easily benefit from the same functionality available to pages, is definitely positive and gives developers flexibility. But we should be cautious about how we sell that as benefit and as an approach for development. Given the option to develop a news app that wasn't tied to a page tree, I probably would have taken it, because that would have felt like the most comfortable fit initially, coming from experiences of other CMSs (even though a small change in thinking would give me something way more flexible!). As with anything, documentation would be hugely important.

**Comment by alexgleason** [comment]
When I first got into Wagtail, I felt quite bewildered and limited by the page structure, when trying to build a very specific site. It was the site for my previous company which is currently back on Craft CMS after not being able to make it work how we wanted in Wagtail. Check out this About page and ponder how it could possibly be implemented in Wagtail: http://pennebaker.com/about

There is no `/team`, by design. We wanted it that way. There's no `/capabilities`, either. Yet, team members and capabilities are displayed on the about page.

If you design all of your sites _around_ Wagtail, you may not have this problem. But that means you're not putting design first, and are choosing to limit yourself with workaround. This will always be a concern, since no CMS can cover literally everything. But this is something that most existing CMS already offer, and what I perceive to be a major caveat of Wagtail. In turn, Wagtail solves a big problem that other CMS's _haven't_ solved with its page explorer. But we need the ability to have both. Sites are designed this way, because sometimes it's what the client wants, and sometimes it makes the most sense. I think this is a perfectly valid design that should be able to be implemented in Wagtail.

Also, many-to-many relationships are an integral part of the sites I've worked on. When you have a IA comprised of `institutions` which can have `team members` who can be the author of `publications` and you need to display each of those elements in a myriad of ways on pages, and `publications` don't have their own pages but need to be displayed in a certain order, nesting them just doesn't work. You'll want to use models. But it's valid to want those models to also be able to be rearranged and version controlled.

I mean, yeah. You can arbitrarily add "pages" to the explorer filled with types of data and then not display them on the front-end. But I imagine that's quite disassociative to the client as well, just presented in a worse way. And having the same thing in two places feels like a hack.

**Comment by ababic** [comment]
@alexgleason I'm with you on pretty much everything you said there. I just wanted to make a case that we shouldn't change things too much... there are lots of things that are right. It's just about finding a balance that will work, I guess.

The lack of many-to-many relationship support is a pain in the ass, which I'm sure the torchbox devs love hearing about :) but hopefully that one is on it's way to being resolved.

wagtailmodeladmin has been limited so far by it's very existence as a separate app, unable to override core functionality, and even as it becomes a contrib app, it's still very much the same thing. It's always been my hope that as part of wagtail, it can grow, become more integrated, and offer an 'in' for solving the kinds of problems you mentioned above.

**Comment by heymonkeyriot** [comment]
Yes, the lack of many-to-many relationship is a pain. #231 is marked on the roadmap for 1.5 release, which will be great to have!

It's also exciting to see the movement on this issue too. Given editors/ developers to organise their content model around the content they have, rather than forcing them in to a tree structure would be incredibly powerful. I think it's the case of giving the flexibility to allow both to exist.

**Comment by tomdyson** [comment]
We're hoping to meet up with @ababic in the next few weeks (between now and 1.5) to discuss the next steps with wagtailmodeladmin.

**Comment by tomusher** [comment]
Glad to see this progressing. I've been working with, and generally enjoying using Wagtail for the past few months but find the whole 'Page' concept a bit counterintuitive so I've just been using wagtailmodeladmin for almost everything (I've gone as far as to remove the Explorer altogether in one site). Maybe that defeats the point of using Wagtail a bit but I think it provides the best content management experience in the Django world at the moment.

To echo @alexgleason's note; I find that generally when architecting websites we've moved to thinking about 'content types/models' and their (complex) relationships as opposed to a tree-based Page structure. The content and the URLs that represent them are generally decoupled.

To me, an ideal future for Wagtail would be one where:

- The concept of Pages is limited to being a tool through which content editors can be their own page designers and build complex pages using StreamField - making use of blocks similar to those in @alexgleason's Craft CMS example to give clients the power to structure individual pages however they want.
- The Explorer becomes a tool for filtering/managing a flat content structure - not a place for managing URLs.
- URLs to models are handled using Django's URL dispatcher at the discretion of the developer, tools for mapping content to URLs are available within Wagtail for content editors to override in limited cases.

It might just be that Wagtail isn't the CMS for me, but I'd love to stick with it and whatever direction you all take, I'm excited to see wagtailmodeladmin become part of core.

**Comment by alexgleason** [comment]
@ababic 
>Given the option to develop a news app that wasn't tied to a page tree, I probably would have taken it ... even though a small change in thinking would give me something much more flexible.

Was rereading this thread and wanted to comment on this.

I feel like the case could be rationally argued for blog posts as channels or pages, depending on how you'd like to implement it. I think this may be something that we'll continue to experiment with and learn from. But a blog post is different from the previously mentioned types, because it's a document.

When you consider the history of the web, it was built to serve pages. Even still HTML is designed around the concept that one HTML file is a document. It has a page body, headers, paragraphs, sections, etc. If you read the official docs by the W3C they have an unhealthy bias towards HTML pages being literal text documents. Like, a page from a book, and not necessarily a piece of software with a UI that the modern web delivers.

So to me, a blog post is synonymous with that. It's tied to the way the web _delivers_ content. I can definitely understand the argument for them as pages from that perspective.

The web was never built to convey the concept of a "team member" through HTTP. Lately people have cleverly been doing it with JSON, which is why I'm excited about building sites with APIs. But these are the real problem points in Wagtail.

So, in short, I don't necessarily think a blog page should be a channel instead of a page. It's probably best left up to the implementer, and is something worth experimenting with.

**Comment by ababic** [comment]
@alexgleason I know it might seem like I missed that point, but I figured those other content types would be easy enough to represent using simple TeamMember or Capability models, hooked in using an InlinePanel on the page to manage them. I was writing on my phone, on the bus, and felt like was already rambling, and wanted to draw to a close. 

But, these workarounds are second nature to me, so providing my solution to that problem is irrelevant to the point, and that is: 

Wagtail could be more flexible and intuitive for developers, who might not know all the workarounds. 

And that's something we're all on the same Page with (sorry, couldn't resist).

**Comment by nek4life** [comment]
When I first started using Wagtail it was a bit confusing to try and turn every model into a set of pages that are in a tree structure. Currently I'm in the process of building an events calendar. Certainly I could make all the events pages, however, the page slugs are almost guarantee to collide with each other due to duplicate slugs being auto generated.

I keep wanting to put the events into a snippet instead, but then I lose most of the page functionality. It would be great to extend more functionality to snippets so you could use them instead of pages where it makes sense.

For me it comes down to how the pages are resolved while keeping a lot of the page functionality like say being able to link back to it with the wagtail toolbar. Tree resolution works well for a lot of things and the routable page mixin works fairly well (although I don't really enjoy putting my routes on my model,) but it would be great to have another page detail resolution type.

**Comment by alexgleason** [comment]
One way we could do this is to allow for multiple explorers.

```python
class PageExplorer(Explorer):
    base_url = '/'
    icon = 'page'

class GrantsExplorer(Explorer):
    base_url = '/grants'
    max_depth = 1
    icon = 'grants'

class Page(models.Model):
    # ...
    allowed_explorers = ['PageExplorer']

class Grant(Page):
    # ...
    allowed_explorers = ['GrantsExplorer']
```

This would create a new explorer menu for "grants."

In the database, a Grant would have a field storing the tree path (just as normal pages do). It would also have an `explorer_id` field which identifies which explorer the path is relative to.

This might not be the best design, just brainstorming.

**Comment by dan-gamble** [comment]
Sorry to bring up an old thread for this, figured it would be better to do that than to start a new issue.

Has there been any ground made on this?

I'm looking in to Wagtail and my only personal stumbling block i've found with Wagtail is that i have to go through the explorer to add any kind of meaningful content, which i find a little inconvenient.

Is there a way to add something along the lines of a `BlogIndexPage` then on the side menu we could have the `Blog` app under which we add articles? You could even do `class Article(Page):` but just have these accessible from the sidebar instead of just in the explorer?

Basically more of a throwback to how Django originally handles apps with their models through the sidebar.

I apologise if i haven't made this that clear, gimme a shout if so and i'll have to try and reframe it.

**Comment by tomdyson** [comment]
> Is there a way to add something along the lines of a BlogIndexPage then on the side menu we could have the Blog app under which we add articles? You could even do class Article(Page): but just have these accessible from the sidebar instead of just in the explorer?

wagtailmodeladmin does exactly this:

http://docs.wagtail.io/en/v1.13.1/reference/contrib/modeladmin/#a-simple-example

**Comment by dan-gamble** [comment]
I have just realised that, should've read the OP more clearly :). Have any of the issues been alleviated since like ordering and versioning?

**Comment by tomdyson** [comment]
> Have any of the issues been alleviated since like ordering and versioning?

Yes, if you're using wagtailmodeladmin to manage Wagtail pages you get all the usual page model features (although AFAIK you would still have to use the Explorer view for ordering).

**Comment by dan-gamble** [comment]
Thanks Tom, is there still active consideration on certain parts of this issue or can it be considered closed with `wagtailmodeladmin`?

**Comment by tomdyson** [comment]
I think it can be closed. Thanks for the prompt @dan-gamble!

**Comment by dan-gamble** [comment]
Thanks for your help Tom :)

**Comment by gasman** [comment]
Allowing _non-page_ models to be versioned is still an open issue (but not one that we're likely to get to in the short term).

**Comment by lb-** [comment]
Update - non-page models (if set up as snippets) now have the ability to be revisioned in the admin as of Wagtail 4.0

https://docs.wagtail.org/en/stable/releases/4.0.html#previews-revisions-and-drafts-for-snippets

**Comment by gasman** [comment]
As far as I can see revisioning was the only thing keeping this issue open, so closing now :-)
